### PR TITLE
Bulk message fetch, deletion, and example clear command.

### PR DESCRIPTION
**Pull request by ZeroSumGame**
This adds in the ability to get messages in bulk from the server.  I've also added in an example "clear" command, which is one of the most common features found in Discord bots.

### CODE REVIEW

File: src/dscord/api/client.d [hunk 1]
```diff
@@ -240,10 +275,27 @@ class APIClient {
   }
 
   /**
+    Returns an array of message IDs for a channel up to limit (max 100),
+    filter with respect to supplied messageID.
+  */
+  Message[] channelsMessagesList(Snowflake chan, uint limit = 50, MessageFilter filter = MessageFilter.BEFORE, Snowflake msg = 0){
+    assert(limit <= 100, "The maximum number of messages that can be deleted at one time is 100.");
```
Comment (b1naryth1ef) [inline src/dscord/api/client.d]: This assert message seems wrong. Should also likely be a fully runtime check, assert gets compiled out in release which seems iffy for required validation stuff like this (e.g. we know the request will fail, might as well not make the HTTP round trip to Discord API servers)

File: examples/src/basic.d [hunk 2]
```diff
@@ -43,6 +43,36 @@ class BasicPlugin : Plugin {
...
+    }
+
+    Message[] messages = this.client.getMessages(event.msg.channelID, limit, event.msg.id);
+
+    //Delete the command message itself
+    event.msg.del();
+
+    if(messages.length > 0){
+      this.client.deleteMessages(event.msg.channelID, messages);
+      auto msg = event.msg.replyf("I deleted %s messages for you.", messages.length);
```
Comment (b1naryth1ef) [inline examples/src/basic.d]:
```d
event.msg.replyf("I deleted %s messages for you.", messages.length).after(3.seconds).del();
```

**Comment by ZeroSumGame** [comment]
@b1naryth1ef  Any word on this?  It's been almost a week.

**Comment by b1naryth1ef** [comment]
Ah sorry I forgot to hit the submit review button.

File: examples/src/basic.d [hunk 3]
```diff
@@ -43,6 +43,36 @@ class BasicPlugin : Plugin {
     event.msg.reply("Pong!");
   }
 
+  //An example command that clears messages in the channel
+  @Command("clear")
+  void onClearMessages(CommandEvent event){
```
Comment (b1naryth1ef) [inline examples/src/basic.d]: Nit; If you could make the bracket spacing line up with everything else (e.g. `) {` that would be awesome.

File: examples/src/basic.d [hunk 4]
```diff
@@ -43,6 +43,36 @@ class BasicPlugin : Plugin {
...
+    uint limit = 100;
+
+    //This command can take an integer argument.
+    if(event.args[].length > 0){
```
Comment (b1naryth1ef) [inline examples/src/basic.d]: Why `event.args[]` and not `event.args`?

File: examples/src/basic.d [hunk 5]
```diff
@@ -43,6 +43,36 @@ class BasicPlugin : Plugin {
...
+      try {
+        limit = std.conv.to!int(event.args[0]);
```
Comment (b1naryth1ef) [inline examples/src/basic.d]: preferred syntax here is just `event.args[0].to!int`

**Review by b1naryth1ef — CHANGES_REQUESTED** [review]
Couple nits and minor changes

**Comment by ZeroSumGame** [comment]
I believe this addresses all the issues.  There are lots of other places that this project could use some exception handling and build-time checks, TBH.

Any idea what is causing TravisCI to be upset with the builds?  Everything is working fine on the newest DMD compiler, so the build script may need updating.

**Review by b1naryth1ef — APPROVED** [review]

**Pull request merged**